recipe/bld.bat

```diff
@@ -0,0 +1,29 @@
+mkdir build
+cd build
+
+cmake %SRC_DIR% -G "Ninja" ^
+      -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+      -DCMAKE_PREFIX_PATH=%LIBRARY_PREFIX% ^
+      -DCMAKE_BUILD_TYPE="Release" ^
+      -DBUILD_PYTHON_BINDINGS=ON ^
+      -DBUILD_CLI_EXECUTABLES=ON ^
+      -DBUILD_SHARED_LIBS=OFF ^
+      -DBUILD_TESTS=OFF ^
+      -DMATLAB_BINDINGS=OFF ^
+      -DDOWNLOAD_STB_IMAGE=OFF ^
+      -DDOWNLOAD_ENSMALLEN=OFF ^
+      -DPYTHON_EXECUTABLE=%PYTHON% ^
+      -DUSE_OPENMP=ON ^
+      -DCMAKE_CXX_COMPILER=cl.exe ^
+      -DCMAKE_C_COMPILER=cl.exe ^
+      -DBoost_INCLUDE_DIRS=%LIBRARY_PREFIX%/include ^
+      -DBoost_USE_STATIC_LIBS=ON
+      -DENSMALLEN_INCLUDE_DIR=%LIBRARY_PREFIX%/include ^
+      -DARMADILLO_INCLUDE_DIR=%LIBRARY_PREFIX%/include ^
+      -DARMADILLO_LIBRARIES="%LIBRARY_PREFIX%/lib/armadillo.lib;%LIBRARY_PREFIX%/lib/openblas.lib" ^
```

Comment:
I'm also surprised, it doesn't look like this changed anything either!  You could run CMake with `--trace` or `--trace-expand`, and maybe we could see what's actually going on.  I can see in the link call of the debug output that it's linking against `armadillo.lib`:

```
2020-05-09T16:33:34.3763765Z LINK: command "C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo src\mlpack\methods\adaboost\CMakeFiles\mlpack_adaboost.dir\adaboost_main.cpp.obj /out:mlpack_adaboost.exe /implib:mlpack_adaboost.lib /pdb:mlpack_adaboost.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console -LIBPATH:%PREFIX%\Library\lib mlpack.lib %PREFIX%\Library\lib\armadillo.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:mlpack_adaboost.exe.manifest" failed (exit code 1120) with the following output:
```

but it's not showing `openblas.lib` in that list.  So I wonder if somehow that's not getting passed down right.

(Boost also isn't included on that list, but I think that's okay---MSVC should be automatically linking against it even though we didn't specify the library name explicitly.)

Reply:
Sorry about those previous commit I didn't know how to use --trace it was a new thing for me.

Reply:
Can you try `--trace-expand`?  It might give us more helpful information about what's going on.  Also, I feel like you already tried `BOOST_ALL_DYN_LINK`, but it wasn't successful.  Personally I'd like to see a log with:

```
      -DBoost_USE_STATIC_LIBS=ON ^
      -DBoost_USE_STATIC_RUNTIME=ON ^
      -DBoost_LIB_DIAGNOSTIC_DEFINITIONS=1 ^
```

and `--trace-expand`; I think that could be helpful in figuring out what's going on. :+1:

Reply:
Yes I did I was just trying to use it with CMAKE_CXX_FLAGS. I thought it might work

Reply:
Added trace-expand and removed `CMAKE_CXX_FLAGS`. I think `-DBoost_USE_STATIC_LIBS=ON ^` is not working.

Reply:
Awesome, thank you for doing that.  I haven't parsed the whole log yet, but I do think I found out what's wrong with Armadillo/OpenBLAS.  First some backstory:

The `armadillo.lib` (or `libarmadillo.so`) library really is just a "wrapper" library.  On Linux (and OS X) systems, since everything is linked dynamically, simply linking against `libarmadillo.so` is sufficient, as when runtime comes around, the runtime linker will then load all dependencies of `libarmadillo.so` (which include OpenBLAS, etc.).

On top of this, Armadillo allows header-only operation, if the `ARMA_USE_WRAPPER` macro is *not* defined (see `/usr/include/armadillo_bits/config.hpp` on your system to see what I mean).  If you don't use `ARMA_USE_WRAPPER`, then (on Linux and OS X) you have to link with `-lopenblas ...` instead of `-larmadillo` (the `...` is for other optional Armadillo dependencies, like ARPACK, hdf5, etc.).

But on Windows, it's a little bit different---in our situation, linking against just `armadillo.lib` is *not* sufficient, even if `ARMA_USE_WRAPPER` is defined.  Windows requires us to link even against the dependencies, so, we must link against `openblas.lib` too.

Now, take a look at `CMake/FindArmadillo.cmake`: on line 82, you can see that we link against only `armadillo.lib` if `ARMA_USE_WRAPPER` is set; but only if it isn't do we add all the other dependencies.

So, I can see from the build logs here that `ARMA_USE_WRAPPER` is set:

```
2020-05-10T17:51:09.7219291Z D:/bld/mlpack_1589132667922/work/CMake/FindArmadillo.cmake(70):  string(REGEX MATCH ARMA_USE_WRAPPER _ARMA_USE_WRAPPER #define ARMA_USE_LAPACK;#define ARMA_USE_BLAS;#define ARMA_USE_NEWARP;#define ARMA_USE_WRAPPER )
```

(The --trace-expand output is weird, but that seems to indicate that `#define ARMA_USE_WRAPPER` does exist in `armadillo_bits/config.hpp`, so I believe the wrapper is enabled.)

That means that `ARMADILLO_LIBRARIES` will contain *only* `armadillo.lib` after CMake is done configuring.  So, we need to find a way to ensure that `ARMADILLO_LIBRARIES` will contain more than that.  If I am reading the CMake file correctly, it *might* be possible to specify the following CMake options and it will work:

```
-DARMADILLO_LIBRARY="%LIBRARY_PREFIX%/lib/armadillo.lib"
-D_ARMA_SUPPORT_LIBRARIES="%LIBRARY_PREFIX%/lib/openblas.lib"
```

and maybe we can see if that works.  The key line we'll be looking for is in the build output, we want to see that `openblas.lib` is in the linker list:

```
2020-05-10T17:54:00.3396291Z cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=src\mlpack\methods\nmf\CMakeFiles\mlpack_nmf.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo src\mlpack\methods\nmf\CMakeFiles\mlpack_nmf.dir\nmf_main.cpp.obj  /out:mlpack_nmf.exe /implib:mlpack_nmf.lib /pdb:mlpack_nmf.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console -LIBPATH:%PREFIX%\Library\lib mlpack.lib  %PREFIX%\Library\lib\armadillo.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
```

(That's the output from this build, way down near the bottom.  If the fix above works, then `openblas.lib` will appear in that list too.)

If that *doesn't* work, maybe we want to try patching `CMake/FindArmadillo.cmake`?  You could try writing a modified `CMake/FindArmadillo.cmake` that checks if the platform is Windows and then finds all the Armadillo support libraries and appends them to `ARMADILLO_LIBRARIES`, instead of the current logic that's based only on `ARMA_USE_WRAPPER`.  Then you could drop that into place during this build, and maybe it would work.

I still haven't addressed the Boost issue at all---I'll look through the logs for that more closely soon, but if we can at least solve the Armadillo issue we'll be getting a good bit closer I think. :+1:

Reply:
Thanks @rcurtin for such a detailed response. I will try your suggestions.

Reply:
:( No dice on the first try:

```
2020-05-11T12:46:45.8289152Z cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=src\mlpack\methods\logistic_regression\CMakeFiles\mlpack_logistic_regression.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo src\mlpack\methods\logistic_regression\CMakeFiles\mlpack_logistic_regression.dir\logistic_regression_main.cpp.obj  /out:mlpack_logistic_regression.exe /implib:mlpack_logistic_regression.lib /pdb:mlpack_logistic_regression.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console -LIBPATH:%PREFIX%\Library\lib mlpack.lib  %PREFIX%\Library\lib\armadillo.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
```

`openblas.lib` isn't in that list.  Maybe you can try modifying `CMake/FindArmadillo.cmake`?

Reply:
Yeah, we need to add a patch.

Reply:
Nice!  It looks like now it is linking against `openblas.lib` correctly:

```
cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=src\mlpack\methods\adaboost\CMakeFiles\mlpack_adaboost.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100177~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo src\mlpack\methods\adaboost\CMakeFiles\mlpack_adaboost.dir\adaboost_main.cpp.obj  /out:mlpack_adaboost.exe /implib:mlpack_adaboost.lib /pdb:mlpack_adaboost.pdb /version:0.0  /machine:x64 /INCREMENTAL:NO /subsystem:console -LIBPATH:%PREFIX%\Library\lib mlpack.lib  %PREFIX%\Library\lib\armadillo.lib  %PREFIX%\Library\lib\openblas.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
```

and now the only linker errors are from Boost:

```
2020-05-11T14:58:53.6662426Z adaboost_main.cpp.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > boost::program_options::arg" (?arg@program_options@boost@@3V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@A) referenced in function "public: virtual class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl boost::program_options::typed_value<int,char>::name(void)const " (?name@?$typed_value@HD@program_options@boost@@UEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
2020-05-11T14:58:53.6664388Z adaboost_main.cpp.obj : error LNK2019: unresolved external symbol "public: static unsigned int const boost::program_options::options_description::m_default_line_length" (?m_default_line_length@options_description@program_options@boost@@2IB) referenced in function "void __cdecl mlpack::bindings::cli::ParseCommandLine(int,char * *)" (?ParseCommandLine@cli@bindings@mlpack@@YAXHPEAPEAD@Z)
2020-05-11T14:58:53.6665484Z mlpack_adaboost.exe : fatal error LNK1120: 2 unresolved externals
```

Looks like we might need to update the conda-forge armadillo package to build in release mode though:

```
2020-05-11T14:58:53.6659101Z armadillo.lib(wrapper.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in adaboost_main.cpp.obj
2020-05-11T14:58:53.6660106Z armadillo.lib(wrapper.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MD_DynamicRelease' in adaboost_main.cpp.obj
```

That seems to imply that `armadillo.lib` was compiled in debug mode.

Reply:
> and now the only linker errors are from Boost:

Yes I was also thinking about Boost I tried to use last example https://cmake.org/cmake/help/v3.15/module/FindBoost.html 
with 
`-DBoost_USE_STATIC_RUNTIME=OFF`
But no changes